File: llm-harness-js/src/stats/data-sources/llm/openrouter-scraper.ts
import { fetchWithTimeout, nowEpochSeconds } from "../utils.js";

const OPENROUTER_MODELS_URL = "https://openrouter.ai/api/frontend/models";
const OPENROUTER_THROUGHPUT_URL =
  "https://openrouter.ai/api/frontend/stats/throughput-comparison";
const OPENROUTER_LATENCY_URL =
  "https://openrouter.ai/api/frontend/stats/latency-comparison";
const OPENROUTER_E2E_LATENCY_URL =
  "https://openrouter.ai/api/frontend/stats/latency-e2e-comparison";
const OPENROUTER_EFFECTIVE_PRICING_URL =
  "https://openrouter.ai/api/frontend/stats/effective-pricing";

const DEFAULT_TIMEOUT_MS = 30_000;
const DEFAULT_CONCURRENCY = 8;
const DEFAULT_MAX_RETRIES = 3;
const DEFAULT_RETRY_BASE_DELAY_MS = 300;

type JsonObject = Record<string, unknown>;

type OpenRouterFrontendModel = {
  slug?: string | null;
  permaslug?: string | null;
};

type OpenRouterStatsPoint = {
  x?: string;
  y?: Record<string, number | null>;
};

type OpenRouterStatsResponse = {
  data?: OpenRouterStatsPoint[];
};

type OpenRouterModelStats = {
  throughput?: OpenRouterStatsResponse | null;
  latency?: OpenRouterStatsResponse | null;
  latency_e2e?: OpenRouterStatsResponse | null;
};

type OpenRouterEffectivePricingResponse = {
  data?: {
    weightedInputPrice?: number | null;
    weightedOutputPrice?: number | null;
  };
};

/**
 * Options for scraping OpenRouter performance stats for a selected model list.
 */
export type OpenRouterScraperOptions = {
  modelIds: string[];
  timeoutMs?: number;
  concurrency?: number;
  maxRetries?: number;
  retryBaseDelayMs?: number;
};

export type OpenRouterPerformanceSummary = {
  throughput_tokens_per_second_median: number | null;
  latency_seconds_median: number | null;
  e2e_latency_seconds_median: number | null;
};

export type OpenRouterPricingSummary = {
  weighted_input_price_per_1m: number | null;
  weighted_output_price_per_1m: number | null;
};

export type OpenRouterScrapedModel = {
  id: string;
  permaslug: string | null;
  performance: OpenRouterPerformanceSummary;
  pricing: OpenRouterPricingSummary;
};

export type OpenRouterScrapedPayload = {
  fetched_at_epoch_seconds: number;
  total_requested_models: number;
  total_resolved_models: number;
  models: OpenRouterScrapedModel[];
};

export type OpenRouterSingleModelOptions = Omit<
  OpenRouterScraperOptions,
  "modelIds"
>;

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

function sanitizeModelId(modelId: string): string {
  return modelId
    .trim()
    .toLowerCase()
    .replace(/:free$/, "");
}

function asRecord(value: unknown): JsonObject {
  return value != null && typeof value === "object" && !Array.isArray(value)
    ? (value as JsonObject)
    : {};
}

function asFiniteNumber(value: unknown): number | null {
  const numericValue = Number(value);
  return Number.isFinite(numericValue) ? numericValue : null;
}

function finiteNumbers(values: unknown[]): number[] {
  return values
    .map((value) => Number(value))
    .filter((value) => Number.isFinite(value));
}

function median(values: number[]): number | null {
  if (values.length === 0) {
    return null;
  }
  const sorted = [...values].sort((left, right) => left - right);
  const mid = Math.floor(sorted.length / 2);
  if (sorted.length % 2 === 1) {
    return sorted[mid] ?? null;
  }
  const left = sorted[mid - 1];
  const right = sorted[mid];
  if (left == null || right == null) {
    return null;
  }
  return (left + right) / 2;
}

function average(values: number[]): number | null {
  if (values.length === 0) {
    return null;
  }
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function toDailyAveragedValues(
  response: OpenRouterStatsResponse | null,
  scaleToSeconds: boolean,
): number[] {
  if (!response || !Array.isArray(response.data)) {
    return [];
  }
  return response.data
    .map((point) => {
      const y = asRecord(point.y);
      const values = finiteNumbers(Object.values(y));
      const dailyAverage = average(values);
      if (dailyAverage == null) {
        return null;
      }
      return scaleToSeconds ? dailyAverage / 1000 : dailyAverage;
    })
    .filter(
      (value): value is number => value != null && Number.isFinite(value),
    );
}

function summarizePerformance(
  stats: OpenRouterModelStats,
): OpenRouterPerformanceSummary {
  const throughputValues = toDailyAveragedValues(
    stats.throughput ?? null,
    false,
  );
  const latencyValues = toDailyAveragedValues(stats.latency ?? null, true);
  const e2eLatencyValues = toDailyAveragedValues(
    stats.latency_e2e ?? null,
    true,
  );

  return {
    throughput_tokens_per_second_median: median(throughputValues),
    latency_seconds_median: median(latencyValues),
    e2e_latency_seconds_median: median(e2eLatencyValues),
  };
}

function summarizePricing(
  response: OpenRouterEffectivePricingResponse | null,
): OpenRouterPricingSummary {
  const data = asRecord(response?.data);
  return {
    weighted_input_price_per_1m: asFiniteNumber(data.weightedInputPrice),
    weighted_output_price_per_1m: asFiniteNumber(data.weightedOutputPrice),
  };
}

function emptyScrapedModel(modelId: string): OpenRouterScrapedModel {
  return {
    id: modelId,
    permaslug: null,
    performance: summarizePerformance({}),
    pricing: summarizePricing(null),
  };
}

async function fetchJsonWithRetry<T>(
  url: string,
  timeoutMs: number,
  maxRetries: number,
  retryBaseDelayMs: number,
): Promise<T> {
  let lastError: unknown = null;

  for (let attempt = 0; attempt < maxRetries; attempt += 1) {
    try {
      const response = await fetchWithTimeout(url, {}, timeoutMs);
      if (!response.ok) {
        const status = response.status;
        if ((status === 429 || status >= 500) && attempt < maxRetries - 1) {
          const backoffMs =
            retryBaseDelayMs * 2 ** attempt + Math.floor(Math.random() * 100);
          await sleep(backoffMs);
          continue;
        }
        throw new Error(`OpenRouter request failed: ${status} (${url})`);
      }
      return (await response.json()) as T;
    } catch (error) {
      lastError = error;
      if (attempt < maxRetries - 1) {
        const backoffMs =
          retryBaseDelayMs * 2 ** attempt + Math.floor(Math.random() * 100);
        await sleep(backoffMs);
        continue;
      }
    }
  }

  throw lastError ?? new Error(`OpenRouter request failed: ${url}`);
}

async function mapWithConcurrency<T, R>(
  items: T[],
  concurrency: number,
  mapper: (item: T, index: number) => Promise<R>,
): Promise<R[]> {
  const safeConcurrency = Math.max(1, Math.floor(concurrency));
  const results = new Array<R>(items.length);
  let cursor = 0;

  async function worker(): Promise<void> {
    for (;;) {
      const index = cursor;
      cursor += 1;
      if (index >= items.length) {
        return;
      }
      results[index] = await mapper(items[index] as T, index);
    }
  }

  await Promise.all(
    Array.from({ length: Math.min(safeConcurrency, items.length) }, () =>
      worker(),
    ),
  );
  return results;
}

function buildPermaslugLookup(
  models: OpenRouterFrontendModel[],
): Map<string, string> {
  const permaslugBySlug = new Map<string, string>();
  for (const model of models) {
    if (typeof model.slug !== "string" || typeof model.permaslug !== "string") {
      continue;
    }
    const slug = sanitizeModelId(model.slug);
    const permaslug = model.permaslug.trim();
    if (!slug || !permaslug) {
      continue;
    }
    permaslugBySlug.set(slug, permaslug);
  }
  return permaslugBySlug;
}

async function fetchPerformanceForPermaslug(
  permaslug: string,
  timeoutMs: number,
  maxRetries: number,
  retryBaseDelayMs: number,
): Promise<{
  performance: OpenRouterModelStats;
  pricing: OpenRouterEffectivePricingResponse;
}> {
  const query = new URLSearchParams({ permaslug });
  const [throughput, latency, latencyE2e, effectivePricing] = await Promise.all(
    [
      fetchJsonWithRetry<OpenRouterStatsResponse>(
        `${OPENROUTER_THROUGHPUT_URL}?${query.toString()}`,
        timeoutMs,
        maxRetries,
        retryBaseDelayMs,
      ),
      fetchJsonWithRetry<OpenRouterStatsResponse>(
        `${OPENROUTER_LATENCY_URL}?${query.toString()}`,
        timeoutMs,
        maxRetries,
        retryBaseDelayMs,
      ),
      fetchJsonWithRetry<OpenRouterStatsResponse>(
        `${OPENROUTER_E2E_LATENCY_URL}?${query.toString()}`,
        timeoutMs,
        maxRetries,
        retryBaseDelayMs,
      ),
      fetchJsonWithRetry<OpenRouterEffectivePricingResponse>(
        `${OPENROUTER_EFFECTIVE_PRICING_URL}?${query.toString()}`,
        timeoutMs,
        maxRetries,
        retryBaseDelayMs,
      ),
    ],
  );

  return {
    performance: {
      throughput,
      latency,
      latency_e2e: latencyE2e,
    },
    pricing: effectivePricing,
  };
}

/**
 * Scrape OpenRouter performance stats for a finalized set of model IDs.
 *
 * This intentionally avoids full-catalog scraping and only fetches stats for
 * `options.modelIds`.
 */
export async function getOpenRouterScrapedStats(
  options: OpenRouterScraperOptions,
): Promise<OpenRouterScrapedPayload> {
  const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const concurrency = options.concurrency ?? DEFAULT_CONCURRENCY;
  const maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
  const retryBaseDelayMs =
    options.retryBaseDelayMs ?? DEFAULT_RETRY_BASE_DELAY_MS;
  const uniqueModelIds = Array.from(
    new Set(options.modelIds.map((modelId) => modelId.trim()).filter(Boolean)),
  );

  const modelDirectory = await fetchJsonWithRetry<{
    data?: OpenRouterFrontendModel[];
  }>(OPENROUTER_MODELS_URL, timeoutMs, maxRetries, retryBaseDelayMs);
  const permaslugBySlug = buildPermaslugLookup(modelDirectory.data ?? []);

  const models = await mapWithConcurrency(
    uniqueModelIds,
    concurrency,
    async (modelId) => {
      const permaslug = permaslugBySlug.get(sanitizeModelId(modelId)) ?? null;
      if (!permaslug) {
        return emptyScrapedModel(modelId);
      }

      try {
        const stats = await fetchPerformanceForPermaslug(
          permaslug,
          timeoutMs,
          maxRetries,
          retryBaseDelayMs,
        );

        return {
          id: modelId,
          permaslug,
          performance: summarizePerformance(stats.performance),
          pricing: summarizePricing(stats.pricing),
        } satisfies OpenRouterScrapedModel;
      } catch {
        // Keep batch resilient: one model failure should not null out the entire payload.
        return {
          id: modelId,
          permaslug,
          performance: summarizePerformance({}),
          pricing: summarizePricing(null),
        } satisfies OpenRouterScrapedModel;
      }
    },
  );

  return {
    fetched_at_epoch_seconds: nowEpochSeconds(),
    total_requested_models: uniqueModelIds.length,
    total_resolved_models: models.filter((model) => model.permaslug != null)
      .length,
    models,
  };
}

/**
 * Fetch OpenRouter performance stats for exactly one OpenRouter model ID.
 *
 * Example input:
 * - `openai/gpt-5.3-codex`
 * - `google/gemini-3.1-pro-preview`
 * - `meta-llama/llama-4-maverick:free` (free suffix is normalized)
 */
export async function getOpenRouterModelStats(
  modelId: string,
  options: OpenRouterSingleModelOptions = {},
): Promise<OpenRouterScrapedModel> {
  const scraperOptions: OpenRouterScraperOptions = {
    modelIds: [modelId],
    ...(options.timeoutMs != null ? { timeoutMs: options.timeoutMs } : {}),
    ...(options.concurrency != null
      ? { concurrency: options.concurrency }
      : {}),
    ...(options.maxRetries != null ? { maxRetries: options.maxRetries } : {}),
    ...(options.retryBaseDelayMs != null
      ? { retryBaseDelayMs: options.retryBaseDelayMs }
      : {}),
  };
  const payload = await getOpenRouterScrapedStats(scraperOptions);

  const firstModel = payload.models[0];
  if (!firstModel) {
    return emptyScrapedModel(modelId);
  }
  return firstModel;
}

File: llm_harness/stats/llm/sources/openrouter_scraper.py
"""Native Python port of the OpenRouter scraper stats source."""

from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor
import math
import random
import re
from statistics import median
import time
from typing import Any, TypedDict
from urllib.parse import urlencode

import httpx

OPENROUTER_MODELS_URL = "https://openrouter.ai/api/frontend/models"
OPENROUTER_THROUGHPUT_URL = "https://openrouter.ai/api/frontend/stats/throughput-comparison"
OPENROUTER_LATENCY_URL = "https://openrouter.ai/api/frontend/stats/latency-comparison"
OPENROUTER_E2E_LATENCY_URL = "https://openrouter.ai/api/frontend/stats/latency-e2e-comparison"
OPENROUTER_EFFECTIVE_PRICING_URL = "https://openrouter.ai/api/frontend/stats/effective-pricing"

DEFAULT_TIMEOUT_MS = 30_000
DEFAULT_CONCURRENCY = 8
DEFAULT_MAX_RETRIES = 3
DEFAULT_RETRY_BASE_DELAY_MS = 300
_ROUTE_SUFFIX_PATTERN = re.compile(r":[a-z0-9._-]+$", re.IGNORECASE)
_SLUG_SPLIT_PATTERN = re.compile(r"[-._/]+")


class OpenRouterScraperOptions(TypedDict, total=False):
    model_ids: list[str]
    timeout_ms: int
    concurrency: int
    max_retries: int
    retry_base_delay_ms: int


class OpenRouterModelOptions(TypedDict, total=False):
    model_id: str
    timeout_ms: int
    concurrency: int
    max_retries: int
    retry_base_delay_ms: int


class OpenRouterPerformanceSummary(TypedDict):
    throughput_tokens_per_second_median: float | None
    latency_seconds_median: float | None
    e2e_latency_seconds_median: float | None


class OpenRouterPricingSummary(TypedDict):
    weighted_input_price_per_1m: float | None
    weighted_output_price_per_1m: float | None


class OpenRouterScrapedModel(TypedDict):
    id: str
    permaslug: str | None
    performance: OpenRouterPerformanceSummary
    pricing: OpenRouterPricingSummary


def _sanitize_model_id(model_id: str) -> str:
    return _ROUTE_SUFFIX_PATTERN.sub("", model_id.strip().lower())


def _as_finite_number(value: object) -> float | None:
    if value is None or isinstance(value, bool):
        return None
    try:
        numeric_value = float(value)
    except (TypeError, ValueError):
        return None
    if not math.isfinite(numeric_value):
        return None
    return numeric_value


def _finite_numbers(values: list[object]) -> list[float]:
    return [numeric_value for value in values if (numeric_value := _as_finite_number(value)) is not None]


def _average(values: list[float]) -> float | None:
    if not values:
        return None
    return sum(values) / len(values)


def _to_daily_averaged_values(
    response: object,
    *,
    scale_to_seconds: bool,
) -> list[float]:
    if not isinstance(response, dict):
        return []
    data = response.get("data")
    if not isinstance(data, list):
        return []

    averaged_values: list[float] = []
    for point in data:
        if not isinstance(point, dict):
            continue
        y = point.get("y")
        if not isinstance(y, dict):
            continue
        values = _finite_numbers(list(y.values()))
        daily_average = _average(values)
        if daily_average is None:
            continue
        averaged_values.append(daily_average / 1000 if scale_to_seconds else daily_average)
    return averaged_values


def _summarize_performance(stats: dict[str, object] | None = None) -> OpenRouterPerformanceSummary:
    safe_stats = stats or {}
    throughput_values = _to_daily_averaged_values(safe_stats.get("throughput"), scale_to_seconds=False)
    latency_values = _to_daily_averaged_values(safe_stats.get("latency"), scale_to_seconds=True)
    e2e_latency_values = _to_daily_averaged_values(safe_stats.get("latency_e2e"), scale_to_seconds=True)
    return {
        "throughput_tokens_per_second_median": median(throughput_values) if throughput_values else None,
        "latency_seconds_median": median(latency_values) if latency_values else None,
        "e2e_latency_seconds_median": median(e2e_latency_values) if e2e_latency_values else None,
    }


def _summarize_pricing(response: object) -> OpenRouterPricingSummary:
    data = response.get("data") if isinstance(response, dict) else None
    safe_data = data if isinstance(data, dict) else {}
    return {
        "weighted_input_price_per_1m": _as_finite_number(safe_data.get("weightedInputPrice")),
        "weighted_output_price_per_1m": _as_finite_number(safe_data.get("weightedOutputPrice")),
    }


def _empty_scraped_model(model_id: str) -> OpenRouterScrapedModel:
    return {
        "id": model_id,
        "permaslug": None,
        "performance": _summarize_performance(),
        "pricing": _summarize_pricing(None),
    }


def _sleep_ms(delay_ms: int) -> None:
    time.sleep(max(0, delay_ms) / 1000)


def _fetch_json_with_retry(
    client: httpx.Client,
    *,
    url: str,
    max_retries: int,
    retry_base_delay_ms: int,
) -> object:
    attempts = max(1, max_retries)
    last_error: Exception | None = None

    for attempt in range(attempts):
        try:
            response = client.get(url)
            if response.status_code == 200:
                return response.json()
            status = response.status_code
            if (status == 429 or status >= 500) and attempt < attempts - 1:
                backoff_ms = retry_base_delay_ms * (2**attempt) + random.randint(0, 99)
                _sleep_ms(backoff_ms)
                continue
            response.raise_for_status()
        except Exception as error:
            last_error = error
            if attempt < attempts - 1:
                backoff_ms = retry_base_delay_ms * (2**attempt) + random.randint(0, 99)
                _sleep_ms(backoff_ms)
                continue
            break

    if last_error is not None:
        raise last_error
    raise RuntimeError(f"OpenRouter request failed: {url}")


def _build_permaslug_lookup(models: object) -> dict[str, str]:
    if not isinstance(models, list):
        return {}

    permaslug_by_slug: dict[str, str] = {}
    for model in models:
        if not isinstance(model, dict):
            continue
        slug = model.get("slug")
        permaslug = model.get("permaslug")
        if not isinstance(slug, str) or not isinstance(permaslug, str):
            continue
        sanitized_slug = _sanitize_model_id(slug)
        trimmed_permaslug = permaslug.strip()
        if not sanitized_slug or not trimmed_permaslug:
            continue
        permaslug_by_slug[sanitized_slug] = trimmed_permaslug
    return permaslug_by_slug


def _has_meaningful_performance(performance: OpenRouterPerformanceSummary) -> bool:
    return (
        performance["throughput_tokens_per_second_median"] is not None or performance["latency_seconds_median"] is not None or performance["e2e_latency_seconds_median"] is not None
    )


def _has_meaningful_pricing(pricing: OpenRouterPricingSummary) -> bool:
    weighted_input = pricing["weighted_input_price_per_1m"]
    weighted_output = pricing["weighted_output_price_per_1m"]
    return (weighted_input is not None and weighted_input > 0) or (weighted_output is not None and weighted_output > 0)


def _split_slug_tokens(value: str) -> list[str]:
    return [token for token in _SLUG_SPLIT_PATTERN.split(value.lower()) if token]


def _token_overlap_score(target_tokens: list[str], candidate_tokens: list[str]) -> float:
    if not target_tokens:
        return 0.0
    target_set = set(target_tokens)
    candidate_set = set(candidate_tokens)
    overlap_count = sum(1 for token in target_set if token in candidate_set)
    return overlap_count / len(target_set)


def _build_slug_fallback_candidates(model_id: str, available_slugs: list[str]) -> list[str]:
    normalized = _sanitize_model_id(model_id)
    if "/" not in normalized:
        return [normalized]

    provider, model_name = normalized.split("/", 1)
    if not provider or not model_name:
        return [normalized]

    target_tokens = _split_slug_tokens(model_name)
    core_prefix = "-".join(target_tokens[:2])
    scored_candidates: list[tuple[str, float, int]] = []
    for slug in available_slugs:
        if not slug.startswith(f"{provider}/") or slug == normalized:
            continue
        candidate_model = slug[len(provider) + 1 :]
        candidate_tokens = _split_slug_tokens(candidate_model)
        overlap_score = _token_overlap_score(target_tokens, candidate_tokens)
        prefix_score = 0.2 if core_prefix and candidate_model.startswith(core_prefix) else 0.0
        score = overlap_score + prefix_score
        if score < 0.6:
            continue
        scored_candidates.append((slug, score, abs(len(candidate_model) - len(model_name))))

    scored_candidates.sort(key=lambda item: (-item[1], item[2], item[0]))
    return [normalized, *[slug for slug, _, _ in scored_candidates[:8]]]


def _resolve_permaslug_candidates(
    model_id: str,
    available_slugs: list[str],
    permaslug_by_slug: dict[str, str],
) -> list[str]:
    resolved_candidates: list[str] = []
    seen: set[str] = set()
    for slug_candidate in _build_slug_fallback_candidates(model_id, available_slugs):
        permaslug = permaslug_by_slug.get(slug_candidate)
        if not permaslug or permaslug in seen:
            continue
        seen.add(permaslug)
        resolved_candidates.append(permaslug)
    return resolved_candidates


def _fetch_performance_for_permaslug(
    permaslug: str,
    *,
    timeout_ms: int,
    max_retries: int,
    retry_base_delay_ms: int,
) -> tuple[dict[str, object], object]:
    query = urlencode({"permaslug": permaslug})
    with httpx.Client(timeout=timeout_ms / 1000) as client:
        throughput = _fetch_json_with_retry(
            client,
            url=f"{OPENROUTER_THROUGHPUT_URL}?{query}",
            max_retries=max_retries,
            retry_base_delay_ms=retry_base_delay_ms,
        )
        latency = _fetch_json_with_retry(
            client,
            url=f"{OPENROUTER_LATENCY_URL}?{query}",
            max_retries=max_retries,
            retry_base_delay_ms=retry_base_delay_ms,
        )
        latency_e2e = _fetch_json_with_retry(
            client,
            url=f"{OPENROUTER_E2E_LATENCY_URL}?{query}",
            max_retries=max_retries,
            retry_base_delay_ms=retry_base_delay_ms,
        )
        effective_pricing = _fetch_json_with_retry(
            client,
            url=f"{OPENROUTER_EFFECTIVE_PRICING_URL}?{query}",
            max_retries=max_retries,
            retry_base_delay_ms=retry_base_delay_ms,
        )

    performance: dict[str, object] = {
        "throughput": throughput,
        "latency": latency,
        "latency_e2e": latency_e2e,
    }
    return performance, effective_pricing


def _fetch_best_available_model_stats(
    model_id: str,
    *,
    available_slugs: list[str],
    permaslug_by_slug: dict[str, str],
    timeout_ms: int,
    max_retries: int,
    retry_base_delay_ms: int,
) -> OpenRouterScrapedModel:
    permaslug_candidates = _resolve_permaslug_candidates(model_id, available_slugs, permaslug_by_slug)
    if not permaslug_candidates:
        return _empty_scraped_model(model_id)

    first_resolved: OpenRouterScrapedModel | None = None
    for permaslug in permaslug_candidates:
        try:
            performance_response, pricing_response = _fetch_performance_for_permaslug(
                permaslug,
                timeout_ms=timeout_ms,
                max_retries=max_retries,
                retry_base_delay_ms=retry_base_delay_ms,
            )
            performance = _summarize_performance(performance_response)
            pricing = _summarize_pricing(pricing_response)
            resolved_model: OpenRouterScrapedModel = {
                "id": model_id,
                "permaslug": permaslug,
                "performance": performance,
                "pricing": pricing,
            }
            if first_resolved is None:
                first_resolved = resolved_model
            if _has_meaningful_performance(performance) or _has_meaningful_pricing(pricing):
                return resolved_model
        except Exception:
            continue

    return first_resolved or _empty_scraped_model(model_id)


def get_openrouter_scraped_stats(
    options: OpenRouterScraperOptions | None = None,
) -> dict[str, Any]:
    options = options or {}
    raw_model_ids = options.get("model_ids") or []
    unique_model_ids = list(dict.fromkeys(model_id.strip() for model_id in raw_model_ids if isinstance(model_id, str) and model_id.strip()))
    timeout_ms = int(options.get("timeout_ms") or DEFAULT_TIMEOUT_MS)
    concurrency = max(1, int(options.get("concurrency") or DEFAULT_CONCURRENCY))
    max_retries = max(1, int(options.get("max_retries") or DEFAULT_MAX_RETRIES))
    retry_base_delay_ms = int(options.get("retry_base_delay_ms") or DEFAULT_RETRY_BASE_DELAY_MS)

    empty_payload = {
        "fetched_at_epoch_seconds": int(time.time()),
        "total_requested_models": len(unique_model_ids),
        "total_resolved_models": 0,
        "models": [_empty_scraped_model(model_id) for model_id in unique_model_ids],
    }
    if not unique_model_ids:
        return empty_payload

    try:
        with httpx.Client(timeout=timeout_ms / 1000) as client:
            model_directory = _fetch_json_with_retry(
                client,
                url=OPENROUTER_MODELS_URL,
                max_retries=max_retries,
                retry_base_delay_ms=retry_base_delay_ms,
            )
        data = model_directory.get("data") if isinstance(model_directory, dict) else None
        permaslug_by_slug = _build_permaslug_lookup(data)
        available_slugs = list(permaslug_by_slug.keys())
        if not available_slugs:
            return empty_payload

        worker_count = min(concurrency, len(unique_model_ids))
        with ThreadPoolExecutor(max_workers=worker_count) as executor:
            models = list(
                executor.map(
                    lambda model_id: _fetch_best_available_model_stats(
                        model_id,
                        available_slugs=available_slugs,
                        permaslug_by_slug=permaslug_by_slug,
                        timeout_ms=timeout_ms,
                        max_retries=max_retries,
                        retry_base_delay_ms=retry_base_delay_ms,
                    ),
                    unique_model_ids,
                )
            )
        return {
            "fetched_at_epoch_seconds": int(time.time()),
            "total_requested_models": len(unique_model_ids),
            "total_resolved_models": sum(1 for model in models if model["permaslug"] is not None),
            "models": models,
        }
    except Exception:
        return empty_payload


def get_openrouter_model_stats(
    options: OpenRouterModelOptions | None = None,
) -> dict[str, Any]:
    model_id = (options or {}).get("model_id")
    if not isinstance(model_id, str):
        return {}

    scraper_options: OpenRouterScraperOptions = {"model_ids": [model_id]}
    if options:
        for option_key in ("timeout_ms", "concurrency", "max_retries", "retry_base_delay_ms"):
            option_value = options.get(option_key)
            if option_value is not None:
                scraper_options[option_key] = option_value

    payload = get_openrouter_scraped_stats(scraper_options)
    models = payload.get("models")
    if isinstance(models, list) and models:
        first_model = models[0]
        if isinstance(first_model, dict):
            return first_model
    return _empty_scraped_model(model_id)


__all__ = [
    "OpenRouterModelOptions",
    "OpenRouterScraperOptions",
    "get_openrouter_model_stats",
    "get_openrouter_scraped_stats",
]
